state/application.go

```diff
@@ -920,6 +941,80 @@ func (a *Application) newCharmStorageOps(
 	return checkStorageOps, upgradeStorageOps, []txn.Op{storageConstraintsOp}, nil
 }
 
+func (a *Application) newCharmDeviceOps(
+	ch *Charm,
+	units []*Unit,
+	updatedDeviceConstraints map[string]devices.Constraints,
+) ([]txn.Op, []txn.Op, []txn.Op, error) {
+
+	fail := func(err error) ([]txn.Op, []txn.Op, []txn.Op, error) {
+		return nil, nil, nil, errors.Trace(err)
+	}
+
+	// Check device to ensure no referenced device is removed, or changed
+	// in an incompatible way. We do this before computing the new device
+	// constraints, as incompatible charm changes will otherwise yield
+	// confusing error messages that would suggest the user has supplied
+	// invalid constraints.
+	sb, err := NewDeviceBackend(a.st)
+	if err != nil {
+		return fail(err)
+	}
+	// oldCharm, _, err := a.Charm()
```

Comment:
we generally don't merge commented out code unless it is intended for immediate use in a follow up patch. I think this is just old code from a copy & paste
.